examples/graphbolt/disk_based_feature/README.md

```diff
@@ -0,0 +1,107 @@
+## Overview
+
+This project demonstrates how to use GraphBolt to train and evaluate a GraphSAGE model for node classification task on large graphs, where node features are on-disk and fetched using `DiskBasedFeature`. GraphBolt utilizes various in-house implemented caching policy algorithms such as [SIEVE](https://cachemon.github.io/SIEVE-website/), [S3-FIFO](https://s3fifo.com), LRU and [CLOCK](https://people.csail.mit.edu/saltzer/Multics/MHP-Saltzer-060508/bookcases/M00s/M0104%20074-12%29.PDF) to cache frequently required features and io_uring to fetch cache-missed features from disk. The SIEVE algorithm is the default option.
+
+# Node classification task
+
+This example demonstrates how to run node classification task with **GraphBolt.DiskBasedFeature**.
+
+## Run on `ogbn-papers100M` dataset
+
+|     Dataset     | Graph Size | Feature Size | Feature Dim |
+| :-------------: | :--------: | :----------: | :---------: |
+| Ogbn-papers100M |   13 GB   |    53 GB    |     128     |
+
+## Results with various caching policies
+
+This part trains a three-layer GraphSAGE model for 3 epochs on `ogbn-papers100M` dataset with 10GB CPU cache, using neighbor sampling.
+
+Below results are collected on an AWS EC2 g5.8xlarge instance with 128GB RAM, 32 cores, an 24GB A10G GPU and a instance storage of 250K IOPS.
+
+### SIEVE
+
+Instruction:
+
+```
+python node_classification.py --gpu-cache-size-in-gigabytes=0 --cpu-cache-size-in-gigabytes=10 --dataset=ogbn-papers100M --epochs=3
+```
+
+Result:
+
+```
+Training: 1178it [04:10,  4.70it/s, num_nodes=665433, gpu_cache_miss=1, cpu_cache_miss=0.0802]
+Evaluating: 123it [00:16,  7.45it/s, num_nodes=625848, gpu_cache_miss=1, cpu_cache_miss=0.0775]
+Epoch 00, Loss: 1.4154, Approx. Train: 0.5776, Approx. Val: 0.6248, Time: 250.58920621871948s
+Training: 1178it [01:56, 10.11it/s, num_nodes=669761, gpu_cache_miss=1, cpu_cache_miss=0.0558]
+Evaluating: 123it [00:14,  8.30it/s, num_nodes=625985, gpu_cache_miss=1, cpu_cache_miss=0.0553]
+Epoch 01, Loss: 1.1435, Approx. Train: 0.6385, Approx. Val: 0.6523, Time: 116.4649817943573s
+Training: 1178it [01:58,  9.93it/s, num_nodes=663480, gpu_cache_miss=1, cpu_cache_miss=0.0478]
+Evaluating: 123it [00:14,  8.56it/s, num_nodes=624204, gpu_cache_miss=1, cpu_cache_miss=0.0476]
+Epoch 02, Loss: 1.0948, Approx. Train: 0.6516, Approx. Val: 0.6590, Time: 118.57369756698608s
+```
+
+### S3-FIFO
```

Comment:
Maybe for the other caching policies, we don't need full output? A simple table showing runtimes for the 3 epochs should be fine? A table of 4 (for each caching policy) by 3 (for each epoch).

Reply:
Results updated. But I'm curious why S3-FiFO seems to perform even worse than LRU? Doesn't it say to outperform LRU in the paper?

Reply:
It does not have to outperform it always. Sometimes other algorithms might perform better. Did you try different cache sizes?

It is also possible I might have introduced a bug, I will run my verification script again.

Reply:
```
import requests
import zstandard as zstd
from struct import iter_unpack
import torch
from dgl.graphbolt.impl import CPUFeatureCache
from cachemonCache import S3FIFO, LRU
from tqdm import tqdm
import os

dtype = torch.int64
num_partitions = 1

base_URL = "https://ftp.pdl.cmu.edu/pub/datasets/twemcacheWorkload/cacheDatasets/"
datasets = [
    (50000, "fiu/fiu_casa-110108-112108.oracleGeneral.zst"),
    (100000, "fiu/fiu_homes-110108-112108.oracleGeneral.zst"),
    (100000, "alibabaBlock/io_traces.ns0.oracleGeneral.zst"),
    (100000, "tencentBlock/tencentBlock.ns1182.oracleGeneral.zst"),
    (1000000, "msr/msr_proj_2.oracleGeneral.zst"),
]
for capacity, dataset in datasets:
    data = requests.get(os.path.join(base_URL, dataset), stream=False).content
    pbar = tqdm(iter_unpack("=IQIq", zstd.ZstdDecompressor().decompress(data)))
    my_caches = {policy: CPUFeatureCache([capacity, 1], dtype, policy, num_partitions) for policy in ["s3-fifo", "lru", "sieve", "clock"]}
    other_caches = {name: cache(capacity) for name, cache in zip(["s3-fifo", "lru"], [S3FIFO, LRU])}
    for i, (timestamp, obj_id, obj_size, next_access_vtime) in enumerate(pbar):
        for cache in my_caches.values():
            values, missing_index, missing_keys, _ = cache.query(torch.tensor([obj_id]))
            if missing_keys.numel() > 0:
                cache.replace(missing_keys, torch.tensor([obj_size]))
        for cache in other_caches.values():
            val = cache.get(obj_id)
            if val is None:
                cache.put(obj_id, obj_size)
        if i % 1000 == 999:
            pbar.set_description(f"My miss rates ({", ".join([f"{name}: {cache.miss_rate:.6f}" for name, cache in my_caches.items()])}), miss rate original: ({", ".join([f"{name}: {1.0 - cache.n_hit / (i + 1):.6f}" for name, cache in other_caches.items()])}), {dataset}")
```

Reply:
Try running this code if you are curious about the performance of different caching on different workload traces.

Reply:
My s3-fifo and reference s3-fifo are almost identical.

Reply:
Also, in my experience, s3-fifo works better with `batch-dependency>1` options compared to others.